File: src/main/java/dev/zprestige/magnesium/mixins/MixinKeyboard.java
package dev.zprestige.magnesium.mixins;

import dev.zprestige.magnesium.Main;
import dev.zprestige.magnesium.event.impl.KeyEvent;
import net.minecraft.client.Keyboard;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

@Mixin(Keyboard.class)
public class MixinKeyboard {

    @Inject(method = "onKey", at = @At( "HEAD"))
    private void onKey(long window, int key, int scancode, int action, int modifiers, CallbackInfo callbackInfo){
        if (key != -1) {
            KeyEvent event = new KeyEvent(key, action);
            Main.Companion.getEventBus().invoke(event);
        }
    }
}

File: src/main/kotlin/dev/zprestige/magnesium/mixins/MixinKeyboard.kt
package dev.zprestige.magnesium.mixins

import dev.zprestige.magnesium.Main
import net.minecraft.client.Keyboard
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo

@Mixin(Keyboard::class)
class MixinKeyboard {

    @Inject(method = ["onKey"], at = [At(value = "HEAD")])
    private fun onKey(window: Long, key: Int, scancode: Int, action: Int, modifiers: Int, info: CallbackInfo) {
        Main.featureManager.keyPressed(key)
    }
}